### ISSUE TITLE
Relax restrictions on designated-storage types

### ISSUE THREAD

**Issue by Bananattack** [wiz-lang/wiz#47]
It is possible to declare a variable as having a "designated storage" type of form `T in storage`.

This means that the variable serves as an alias for the register/global that is listed in the type signature.

eg.
```
var count : u8 in x; // x is a u8 register.
count = 10; // same as x = 10;
```

This is useful for giving descriptive names to locals, and as a mechanism to explicitly pass arguments/return values from functions through registers.

However, it is not currently possible to designate storage unless they match exactly in type. This results in a lot of casts on assignment to/from desigated storage, especially when used as arguments.

```
var ptr1 : *u8 in hl; // error.
var ptr2 : u16 in hl; // ok. but now using it as a pointer is no easier than using the register directly...
```

Registers are typically defined as unsigned integers, to give them some level of type safety in their general usage. But registers are also used as pointers, and doing so requires a cast. If it were possible to declare designated storage locals that reinterpret the bits of the register as a pointer, that would be very handy.

eg. This should be allowed:

```
var ptr : *u8 in hl;
ptr = &mem;
*ptr++ = a; // assuming byte-sized post-increment indirect exists.
(*ptr)++; // inc [hl]
```

much less tedious to write and nicer to read than
```
hl = &mem as u16;
*(hl++ as *u8) = a;
(*(hl as *u8))++;
```

if `u16` and `*u8` are the same size, it should be possible to use either for designated storage.

likewise with for `u8` and `i8`

basically, any type where its size matches the size of the designated storage should be possible.

The tricky detail is maybe the implied casts that happen any time the designated storage register is accessed. It is a cast that converts type but preserves the l-value-ness of the source expression.

Now I'm just braindumping stuff and not fully thinking things through with much rigour because I lost my train of thought. Maybe designated storage is better thought of as a C++ style reference except with a single possible target and no implied pointers behind the scenes. It's essentially kinda like having a `U&` reference that implicitly reinterpret-casts an l-value from its the source type `T&` to the dest type `U&`. or something??? idkh ajfjgkqhfk
